ray-operator/controllers/ray/rayservice_controller.go

```diff
@@ -630,53 +629,86 @@ func (r *RayServiceReconciler) checkIfNeedSubmitServeDeployment(rayServiceInstan
-	if !isServeConfig {
-		shouldUpdate = true
-		reason = fmt.Sprintf("No Serve config has been cached for cluster %s with key %s", rayClusterInstance.Name, cacheKey)
-	} else if !utils.CompareJsonStruct(cachedServeConfig, rayServiceInstance.Spec.ServeDeploymentGraphSpec) {
-		shouldUpdate = true
-		reason = fmt.Sprintf("Current Serve config doesn't match cached Serve config for cluster %s with key %s", rayClusterInstance.Name, cacheKey)
+	serveConfigVersion := r.determineServeConfigVersion(rayServiceInstance)
+	if serveConfigVersion == utils.RayServiceServeConfigV1 {
+		cachedServeConfig, isServeConfig := cachedConfigObj.(rayv1alpha1.ServeDeploymentGraphSpec)
+		if !isServeConfig {
+			shouldUpdate = true
+			reason = fmt.Sprintf("No Serve config has been cached for cluster %s with key %s", rayClusterInstance.Name, cacheKey)
+		} else if !utils.CompareJsonStruct(cachedServeConfig, rayServiceInstance.Spec.ServeDeploymentGraphSpec) {
+			shouldUpdate = true
+			reason = fmt.Sprintf("Current Serve config doesn't match cached Serve config for cluster %s with key %s", rayClusterInstance.Name, cacheKey)
+		}
+		r.Log.V(1).Info("shouldUpdate", "shouldUpdateServe", shouldUpdate, "reason", reason, "cachedServeConfig", cachedServeConfig, "current Serve config", rayServiceInstance.Spec.ServeDeploymentGraphSpec)
+	} else if serveConfigVersion == utils.RayServiceServeConfigV2 {
+		cachedServeConfigV2, isServeConfigV2 := cachedConfigObj.(string)
+		if !isServeConfigV2 {
+			shouldUpdate = true
+			reason = fmt.Sprintf("No Serve config has been cached for cluster %s with key %s", rayClusterInstance.Name, cacheKey)
+		} else if cachedServeConfigV2 != rayServiceInstance.Spec.ServeConfigV2 {
+			shouldUpdate = true
+			reason = fmt.Sprintf("Current Serve config doesn't match cached Serve config for cluster %s with key %s", rayClusterInstance.Name, cacheKey)
+		}
+		r.Log.V(1).Info("shouldUpdate", "shouldUpdateServe", shouldUpdate, "reason", reason, "cachedServeConfig", cachedServeConfigV2, "current Serve config", rayServiceInstance.Spec.ServeConfigV2)
 	}
 
-	r.Log.V(1).Info("shouldUpdate", "shouldUpdateServe", shouldUpdate, "reason", reason, "cachedServeConfig", cachedServeConfig, "current Serve config", rayServiceInstance.Spec.ServeDeploymentGraphSpec)
-
 	return shouldUpdate
 }
 
-// Determines the serve config type from a ray service instance
-// If the user has set a value for `ServeConfigV2`, the config type is MULTI_APP
-// Otherwise, the user should have set a value for `ServeConfig`, in which case the config type is SINGLE_APP
-func (r *RayServiceReconciler) determineServeConfigType(ctx context.Context, rayServiceInstance *rayv1alpha1.RayService) utils.RayServeConfigType {
-	if rayServiceInstance.Spec.ServeConfigV2 == (rayv1alpha1.ServeConfigV2{}) {
-		return utils.SINGLE_APP
+func (r *RayServiceReconciler) determineServeConfigVersion(rayServiceInstance *rayv1alpha1.RayService) utils.RayServiceServeConfigVersion {
```

Comment:
Why do we need both `determineServeConfigVersion` and `ServeConfigType` concurrently? If we choose to have these two together, what is their distinction and relationship? For instance, we could empower users by consistently honoring the value of `ServeConfigType` set by users and only check `rayServiceInstance.Spec.ServeConfigV2 == ""` if `ServeConfigType` is not specified.

Reply:
Now that ServeConfigV2 is just a plain string, ServeConfigV2 can be either single app or multi app

Reply:
determineServeConfigVersion is just an internal thing. It's just so that I don't have to consistently type out the check `rayServiceInstance.Spec.ServeConfigV2 == ""`.

Reply:
Synced offline, serveConfigV2 will only support multi app